developer_manual/app_development/bootstrap.rst

```diff
@@ -44,15 +49,85 @@ An `Application` class shall serve as central initialization point of an app.
 
     namespace OCA\MyApp\AppInfo;
 
+    use OCA\MyApp\Listeners\UserDeletedListener;
+    use OCA\MyApp\Notifications\Notifier;
     use OCP\AppFramework\App;
+    use OCP\AppFramework\Bootstrap\IBootstrap;
+    use OCP\Notification\IManager;
+    use OCP\User\Events;
 
-    class Application extends App {
+    class Application extends App implements IBootstrap {
 
-        public function registerHooks(): void {
-            \OCP\Util::connectHook('OC_User', 'pre_deleteUser', 'OCA\MyApp\Hooks\User', 'deleteUser');
+        public function register(IRegistrationContext $context): void {
+            // ... registration logic goes here ...
+
+            $context->registerEventListener(
+                BeforeUserDeletedEvent::class,
+                UserDeletedListener::class
+            );
+        }
+
+        public function boot(IBootContext $context): void {
+            // ... boot logic goes here ...
+
+            /** @var IManager $manager */
+            $manager = $context->getAppContainer()->query(IManager::class)
+            $manager->registerNotifierService(Notifier::class);
         }
 
     }
 
-.. note:: Nextcloud does not load this class for every request. You should query an instance inside your :ref:`app-php` to
-          load for every request, if desired.
+Bootstrapping process
+---------------------
+
+To give a better overview of *when* each of the bootstrapping stages are reached and how they app can interact with them,
+this section explains the changes done for Nextcloud 20.
+
+Nextcloud 20 and later
+**********************
+
+Nextcloud 20 is the first release with the interface ``\OCP\AppFramework\Bootstrap\IBootstrap``. This interface can be
+implemented by apps' Application class to signal that they want to act on the bootstrapping stages. The major difference
+between this and the old process is that the boostrapping is not performed in a sequence, but apps register and boot
+interleaved. This should ensure that an app that boots can rely on all other apps' registration to be finished.
+
+The overall process is as follows.
+
+1) Each installed and enabled app that has an ``Application`` class class that implements ``IBootstrap``, the ``register``
+   method will be called. This method receives a context argument via which the app can prime the dependency injection
```

Comment:
... a context argument to prepare the dependency injection container and register other services lazily.